### ISSUE TITLE
Huge Changes Incoming

### ISSUE THREAD

**Issue by Linaqruf** [Linaqruf/kohya-trainer#78]
I am working on updating Kohya's script to the latest version. The current version is from February 3rd, but the latest version in sd-scripts is from February 11th. This update will bring many new arguments and changes. I am also restructuring the notebooks to make the code more readable, maintainable, and user-friendly.

Here are some of the changes I have made:

1. I have switched from using git-large-textcaps for auto captioning to using blip captioning with beam search enabled. The reason being that git is slower, while blip with beam search has similar quality.
2. . I have deleted the Image Upscaler with R-ESRGAN cell because the bucketing already upscales images when converting to latents.
3.  I have combined the Login to Huggingface Hub cell with the cell where you define the model and dataset repository, making the workflow easier.
4. New Arguments (useful for advanced users, but may confuse end-users):

- bucket_reso_steps = 64 #@param {type:"slider", min:0, max:100, step:8}
- bucket_no_upscale = False #@param{type:"boolean"}
- caption_dropout_rate = 0 #@param {type:"number"}
- caption_dropout_every_n_epochs = 0 #@param {type:"number"}
- tag_dropout_rate = 0 #@param {type:"number"}

5.I have chosen not to use these and let advanced users register new arguments in additional_arguments.
6.  have combined the data cleaning cell with the RGB -> RGBA converter.
7. I have improved the model converter cell (in both the main and dreambooth notebooks).

Here's a sneak peek:
https://colab.research.google.com/github/Linaqruf/kohya-trainer/blob/8c52bc700f8cbb68a6ee58e4f73e27067d885e5d/kohya-LoRA-dreambooth.ipynb

![image](https://user-images.githubusercontent.com/50163983/218369702-0671ea65-20f4-4008-a0ce-bf204f8a2b51.png)

**Comment by Linaqruf** [comment]
you need to input `dev` to branch field
![image](https://user-images.githubusercontent.com/50163983/218532245-e5c55266-9484-41d1-92f0-5249e502633f.png)

**Comment by Linaqruf** [comment]
> It would be helpful if you can write on there a little about the function of those arguments.

you can read it here, https://github.com/kohya-ss/sd-scripts

**Comment by shinkarom** [comment]
Can you add unet scale and text encoder scale? According to https://github.com/cloneofsimo/lora/discussions/37,in original repo both items can be changed separately, and not having them at 1.0 can sometimes be better.

**Comment by noriwaru** [comment]
When can we expect an update?　Are you busy?

**Comment by GeneralAwareness** [comment]
What I am wanting is the --noise and the --lion kohya has in his script.  Those are big (especially that last one).

**Comment by Linaqruf** [comment]
##### v13 (25/02):
__What Changes?__
- Of course refactoring, cleaning and make the code and cells more readable and easy to maintain.
  - Moved `Login to Huggingface Hub` to `Deployment` section, in the same cell with defining repo.
  - Merged `Install Kohya Trainer`, `Install Dependencies`, and `Mount Drive` cells
  - Merged `Dataset Cleaning` and `Convert RGB to RGBA` cells
  - Deleted `Image Upscaler` cell, because bucketing automatically upscale your dataset (converted to image latents) to `min_bucket_reso` value.
  - Deleted `Colab Ram Patch` because now you can set `--lowvram` in the training script.
  - Revamped `Unzip dataset` cell to make it look simpler
- Added xformers pre-compiled wheel for `A100` 
- Revamped `Pretrained Model` section
  - Deleted some old pretrained model
  - Added `Anything V3.3`, `Chilloutmix`, and `Counterfeit V2.5` as new pretrained model for SD V1.x based model
  - Added `Replicant V1.0`, `WD 1.5 Beta` and `Illuminati Diffusion V1` as new pretrained model for SD V2.x 768v based model
  - Changed `Stable Diffusion 1.5` pretrained model to pruned one.
- Changed Natural Language Captioning back from GIT to BLIP with `beam_search` enabled by default
- Revamped Image Scraper from simple to advanced, added new feature such as:
  - Added `safebooru` to booru list
  - Added `custom_url` option, so you can copy and paste the url instead of specify which booru sites and tags to scrape
  - Added `user_agent` field, because you can't access some image board with default user_agent
  - Added `limit_rate` field to limit your count
  - [Experimental] Added `with_aria2c` to scrape your dataset, not a wrapper, just a simple trick to extract urls with `gallery-dl` and download them with aria2c instead. Fast but seems igonoring `--write-tags`.
  - All downloaded tags now saved with `.txt` format instead of `.jpg.txt`
  - Added `additional_arguments` to make it more flexible if you want to try other args
- Revamped `Append Custom Tag` cell
  - Create new caption file for every image file based on extension provided (`.txt/.caption`) if you didn't want to use BLIP or WD Tagger
  - Added `--keep_tokens` args to the cell
- Revamped `Training Model` section. 
  - Revamped `prettytable` for easier maintenance and bug fixing
  - Now it has 4 major cell:
    - Folder Config
      - To specify `v2`, `v2_parameterization` and all important folder and project_name
    - LoRA and Optimizer Config
      - Only `Optimizer Config` for notebook outside LoRA training
      - All about Optimizer, `learning_rate` and `lr_scheduler` goes here
      - Added new Optimizer from latest kohya-ss/sd-script, all available optimizer : `"AdamW", "AdamW8bit", "Lion", "SGDNesterov", "SGDNesterov8bit", "DAdaptation", "AdaFactor"
      - Currently you can't use `DAdaptation` if you're in Colab free tier because it need more VRAM
      - Added `--optimizer_args` for custom args, useful if you want to try adjusting weight decay, betas etc
     - Dataset Config
      - Only available for Dreambooth method notebook, it basically bucketing cell for Dreambooth.
      - Added `caption dropout`, you can drop your caption or tags by adjusting dropout rates.
      - Added `--bucket_reso_steps` and `--bucket_no_upscale`
     - Training Config
      - Added `--noise_offset`, read [Diffusion With Offset Noise](https://www.crosslabs.org//blog/diffusion-with-offset-noise)
      - Added `--lowram` to load the model in VRAM instead of CPU
- Revamped `Convert Diffusers to Checkpoint` cell, now it's more readable.
- Fixing bugs when `output_dir` located in google drive, it assert an error because of something like `/content/drive/dreambooth_cmd.yaml` which is forbidden, now instead of saved to `{output_dir}`, now training args history are saved to `{training_dir}`

__News__
- I'm in burnout phase, so I'm sorry for the lame update.
- [Fast Kohya Trainer](https://github.com/Linaqruf/kohya-trainer/blob/main/fast-kohya-trainer.ipynb), an idea to merge all Kohya's training script into one cell. Please check it [here](https://colab.research.google.com/github/Linaqruf/kohya-trainer/blob/main/fast-kohya-trainer.ipynb). 
  - Please don't expect high, it just a secondary project and maintaining 1-click cell is hard. So I won't prioritized it.
- Kohya Textual Inversion are cancelled for now, because maintaining 4 Colab Notebook already making me this tired. 
  - Please use this instead, not kohya script but everyone on WD server using this since last year:
    - [stable-textual-inversion-cafe colab](https://colab.research.google.com/drive/1bbtGmH0XfQWzVKROhiIP8x5EAv6XuohJ) 
    - [stable-textual-inversion-cafe Colab - Lazy Edition](https://colab.research.google.com/drive/1ouTImTpYkBrX5hiVWrzJeFtOyKES92uV)
- I wrote a Colab Notebook for #AUTOMATIC1111's #stablediffusion Web UI, with built-in Mikubill's #ControlNet extension. All Annotator and extracted ControlNet model are provided in the notebook. It's called [Cagliostro Colab UI](https://colab.research.google.com/github/Linaqruf/sd-notebook-collection/blob/main/cagliostro-colab-ui.ipynb). Please try it.
  - You can use new UI/UX from [Anapnoe](https://github.com/anapnoe/stable-diffusion-webui) in the notebook. You can find the option in `experimental` section.
![image](https://user-images.githubusercontent.com/50163983/221345472-0f9fd9b3-0fe6-4d9d-af0b-05d9e6001e03.png)

**Comment by Linaqruf** [comment]
Please let me know if there is bugs, and error.

**Comment by GeneralAwareness** [comment]
Will, do.  Big stuff.

**Comment by younyokel** [comment]
pretty confusing fr

**Comment by Linaqruf** [comment]
Ikr, the trainer is more advanced than before, you can use old commit notebook or just train with default value if you find it too hard.

All new optimizer types, args, offset noise, caption dropout, new bucketing option, etc, too much new args to try, but not much time to try all of them.

**Comment by GeneralAwareness** [comment]
I am trying to figure out this lora finetune stuff as I have been using the lora DB so json?  Which one is the best way to go?

**Comment by DarkAlchy** [comment]
I had some time tonight to play with this and I found a bug plus a HUGE question.

The bug:

![image](https://user-images.githubusercontent.com/10561690/221754921-2fb62a38-406e-4785-8525-d7c1e30ab6d9.png)

I have it checked so it should not be saving there.

The other issue is that this now is taking twice as long to do the same training, using the exact same EVERYTHING, as previous.  Why?

A subset is I despise the max epoch/step setting as that is a misnomer as it isn't really a max it sets that to that (overrides).  Max would mean if I do something stupid it will not go over that max not that if I do epochs 1, or the maths says steps are 10 and I set max to 2000 it will do 2000 steps, or in the last case 2000 epochs.  I tried it at -1 thinking that would ignore it but no it just instantly finished at 0.

**Comment by Linaqruf** [comment]
I didn't change anything about default value nor backend script (except tag reading) compared to the previous version. You need to provide a screenshot, training logs, or copied notebook with the same hyperparams and dataset so I can figure out what is the error.

And yes, `max_train_epochs` is not that accurate. 

You can complain about backend script here instead : https://github.com/kohya-ss/sd-scripts

**Comment by GeneralAwareness** [comment]
I noticed this is now taking twice as long as well for the same number of steps myself.  I am not sure why.  I imagine it is something Kohya did tbh.

**Comment by Linaqruf** [comment]
something not passed to the accelerate, 1680 was the default max_steps_

i think this is because the linebreak `\` in the training config
![image](https://user-images.githubusercontent.com/50163983/221766675-baa02e5e-917b-49a4-a9e3-6b5a99b0835f.png)

can you check the train.sh to see all hyperparams passed?

**Comment by Linaqruf** [comment]
i need to change this to yaml asap

**Comment by GeneralAwareness** [comment]
My max steps was 1680, as they should have been and always were but when I told it 4 epochs in the maths it told me to get lost and divided it by 8.

**Comment by GeneralAwareness** [comment]
I ran the old one just now

  num train images * repeats / 学習画像の数×繰り返し回数: 450
  num reg images / 正則化画像の数: 0
  num batches per epoch / 1epochのバッチ数: 150
  num epochs / epoch数: 4
  batch size per device / バッチサイズ: 3
  total train batch size (with parallel & distributed & accumulation) / 総バッチサイズ（並列学習、勾配合計含む）: 3
  gradient accumulation steps / 勾配を合計するステップ数 = 1
  total optimization steps / 学習ステップ数: 600
  
  Now let me run the new one (btw, I really dislike this max thing and much prefer the old method).

**Comment by GeneralAwareness** [comment]
num train images * repeats / 学習画像の数×繰り返し回数: 450
  num reg images / 正則化画像の数: 0
  num batches per epoch / 1epochのバッチ数: 150
  num epochs / epoch数: 4
  batch size per device / バッチサイズ: 3
  total train batch size (with parallel & distributed & accumulation) / 総バッチサイズ（並列学習、勾配合計含む）: 3
  gradient accumulation steps / 勾配を合計するステップ数 = 1
  total optimization steps / 学習ステップ数: 600

Got it.  Line break was part and I had to add code to set the max steps.  Telling ya that max stuff eliminates the entire need of the maths as it break it.

**Comment by younyokel** [comment]
ChillOutMix was taken down @Linaqruf
